File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let cur = head;
    let prev = null;
    let length = 0;
    while (cur != null) {
        cur = cur.next;
        length++;
    }
    cur = head;
    let cnt = 0;
    while (cnt < length-n) {
        prev = cur;
        cur = cur.next;
        cnt++;
    }
    if (length != n) {
        prev.next = cur.next;
    }
    else if (prev != null) {
        head = prev.next;
    }
    else {
        head = cur.next;
    }
    return head;
};

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode sentinel = new ListNode();
        sentinel.next = head;
        ListNode cur = sentinel;
        int len = -1;
        while (cur != null) {
            len++;
            cur = cur.next;
        }
        cur = sentinel;
        for (int i=0; i<len-n; i++) {
            cur = cur.next;
        }
        if (cur.next != null) {
            cur.next = cur.next.next;
        }
        return sentinel.next;
    }
}